core/src/main/scala/org/apache/spark/deploy/master/SchedulingAlgorithm.scala

```diff
@@ -0,0 +1,171 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.master
+
+/**
+ * An interface for sort algorithm
+ */
+private[spark] trait SchedulingAlgorithm {
```

Comment:
I'm not sure if it is the right time to have such an interface, considering we do not have other algorithms except FIFO in standalone mode. 

The current interface is built up based on the current implementation, as a result, some cases might be ignored. As an example, the current algorithm treats `memory` as the first-class citizen, so whether the executors are to be launched is firstly determined by whether there is enough memory. Considering that if we want to implement something like Dominant Resource Fair ( https://www.cs.berkeley.edu/~alig/papers/drf.pdf), then the current interface like `allocateWorkerResourceToExecutors` brings too many limitations here

Let me bring other guys who always maintain these code blocks, @JoshRosen @andrewor14 @srowen